### ISSUE TITLE
Premature thread interruption when using observeOn operator

### ISSUE THREAD

**Issue by visualage** [ReactiveX/RxJava#5711]
When using RxJava 2.1.3 to 2.1.6, we observed a possible racing condition which caused the observeOn thread to be interrupted prematurely.

A sample code snippet could be the following (not a complete example):
```java
Observable.fromPublisher(publisher)
.observeOn(Schedulers.computation())
.firstElement()
.map(operation)
.blockingGet();
```

In the above example, observeOn operator would create a ObservableObserveOn instance. When there is the first item coming (ObservableObserveOn.ObserveOnObserver.onNext() is called), a Runnable is scheduled via the NewThreadWorker.scheduleActual(). Inside NewThreadWorker.scheduleActual(), a ScheduledRunnable is created wrapping the given Runnable, and submit the ScheduledRunnable to the executor for execution. Then the future returned by the executor is set into ScheduledRunnable via setFuture method.

The problem lies when the ScheduledRunnable.dispose() method is called between submitting to the executor and ScheduledRunnable.setFuture method by the underlying Runnable wrapped in ScheduledRunnable (which is executed in parallel by the executor). In this case, the Runnable contains the firstElement() operator would create a ObservableElementAtMaybe instance. ObservableElementAtMaybe.onNext() method would call the dispose() method before calling actual.onSuccess(). So combining these, if ScheduledRunnable.setFuture is called after ScheduledRunnable.dispose() in ObservableElementAtMaybe.onNext(), the actual.onSuccess() would be interrupted, which is not what we want.

**tl;dr** In the sample code, if the execution order is the following, it would create a thread interruption that interfere with the processing of the chained operators.
```
Thread 1: future = executor.submit(ScheduledRunnable)
Thread 2 (executor thread): ScheduledRunnable.dispose()
Thread 1: ScheduledRunnable.setFuture(future)
Thread 2: actual callback (e.g. actual.onSuccess()) <-- Thread interrupted. Throws exception when we call some synchronization methods in a non-blocking way (e.g. Semaphore.tryAcquire(0, TimeUnit.SECONDS)). 
```

A possible solution is to change the order, i.e. call actual.onSuccess() before dispose(). I searched the code base and found out that in general, dispose() is called before the actual callback. I am not sure if there are other concerns with this specific ordering (i.e. dispose before actual callback), but it certainly sounds odd to me.

**Comment by akarnokd** [comment]
`dispose` happens before the `onNext` or `onSuccess` because these may take an arbitrary long time to execute during which the source still doesn't know it should stop producing items. This can lead to excessive memory usage with `Observable`.

Change the order of operators so the `firstElement` doesn't cancel the task performing the thread switching:

```java
Observable.fromPublisher(publisher)
.firstElement()
.observeOn(Schedulers.computation())
.map(operation)
.blockingGet();
```

**Comment by visualage** [comment]
While this can be a workaround for this simple example, but in general, this would be a problem. Here may be a more complex example that is difficult to workaround with:
```
Observable.fromPublisher(publisher)
.firstElement()
.observeOn(Schedulers.computation())
.flatMapObservable(operation)
.firstElement()
.flatMapObservable(operation2)
.observeOn(Schedulers.io())
.map(operation3)
.blockingGet()
```

Another possible solution is in the ScheduledRunnable.run(), wait until the setFuture is called before proceeding to the actual.run(). Again, I am not sure if there is any side-effect on this.

**Comment by akarnokd** [comment]
RxJava aims for non-blocking operations so it can't wait for `setFuture` to be executed. So either you don't call interrupt-sensitive code or create a custom scheduler that doesn't interrupt on cancelling the `Future`.

**Comment by visualage** [comment]
I would argue that the behavior of interrupting a downstream processing thread for observeOn when the observeOn is disposed is questionable. I think it makes perfect sense to interrupt the upstream threads when an operator is disposed, but definitely not downstream threads.

If your argument stands, i.e. we should not have interrupt-sensitive code in RxJava callbacks, why cancel the future with interrupt in the first place?

**Comment by akarnokd** [comment]
> interrupt the upstream threads when an operator is disposed, but definitely not downstream threads

Cancellation of the upstream triggers cleanup which cancels outstanding tasks of the `Worker`. It can't know you still want to keep using that thread and it can't know what cancelled it for what purpose either.

There were experiments that scheduled the cancellation of the worker behind the currently running task but that led to extended worker lifecycle, especially with `Schedulers.io()`, and leaked threads, for example, because the task to be stopped would block afterwards.

The `ExecutorService` API **returns** the `Future` that represents the task which could be already running concurrently, thus `setFuture` is racing with the completion of the task or a `Worker`-wide cancellation. The `ScheduledRunnable` has to use a deferred cancellation schema to get around this late `Future`. I don't know any means to get a "cancellation token" upfront for a task before it is even able to run to avoid that race.

> why cancel the future with interrupt in the first place?

To interrupt potentially blocking code that runs in operators such as `map` or `doOnNext`.

**Comment by visualage** [comment]
Let's put aside the current `Worker` and `ScheduledRunnable` implementation, and think about what's the desired behavior for `observeOn` and the desired behavior of `dispose`. 

To my understanding, `dispose` means notifying all upstream operations including the initial publisher that we no longer want to receive more items. If this understanding is correct, the `observeOn` thread should not be disposed unless the `dispose` of the last operation in the `observeOn` thread is called. E.g.
```
Observable
.create()
.observeOn(computation)
.firstElement()
.map()
.observeOn(io)
.flatMap()
```
The first `observeOn` thread should only be canceled when the `dispose` of `map` operation is called, not when the `dispose` of `firstElement` is called. A possible easier implementation to achieve this would be the `dispose` method of the second `observeOn` would cancel the thread of the first `observeOn`.

**Comment by akarnokd** [comment]
Can't work that way. 

As I said, you need a custom scheduler that doesn't interrupt when its tasks get disposed.

**Comment by visualage** [comment]
Can you elaborate why RxJava's dispose should not work that way, conceptually speaking? I am not saying the current implementation is doing this or can be easily modified to do that.

**Comment by akarnokd** [comment]
> first observeOn thread should only be canceled when the dispose of map operation is called

There is nothing that would call that dispose because it is `firstElement` that disposes its source before map gets executed.

In addition, RxJava operators are independent and standalone in the sense that they know practically nothing about what their upstream or downstream is or doing, only responding to the standard Observer and Disposable method calls they can interact with.

**Comment by adrian-baker** [comment]
>you don't call interrupt-sensitive code

This really isn't feasible, many downstream libraries outside of the callers control are rightly senstive to interrupt calls.

Debugging issues caused by this behaviour is extremely hard. Once traced back to RxJava, and traced back to issues like this and many others, which are essentially dismissed, makes it very hard to continue using the library.

**Comment by akarnokd** [comment]
@adrian-baker You missed the other workaround I suggested: "create a custom scheduler that doesn't interrupt on cancelling the `Future`".

> issues like this and many others, which are essentially dismissed,makes it very hard to continue using the library

I do the maintenance for free and I look at people's profiles to see if they or their employer do significant contribution to open-source. If they do, I'm more inclined to go beyond high level suggestion and contribute the fixes/workarounds myself to their project.

For anybody else, they have to do it by themselves or are [welcome](https://twitter.com/SwiftOnSecurity/status/1067682759592869889) to open a support contract with my employer.

**Comment by ysyyork** [comment]
Just wanna clarify on this. I have a case like below:
```
Single.create(...)
.observeOn(Schedulers.io())
.flatMap(Single.create(...))
.flatMap(Single.create(...))
...
```
Basically it's a change of Singles and the each single is calling onSuccess or onError. In that case, for the second single, no matter if it's calling onSuccess or onError, it means completing the task right? Then if I don't have another observeOn following it, the third single can fail because of interruption? If that's the case, does it mean when chaining singles we need to call observeOn for each single? Thanks.

**Comment by adrian-baker** [comment]
> I do the maintenance for free and I look at people's profiles to see if they or their employer do significant contribution to open-source

I'm not making any judgement at all on your personal contribution. Just trying making the point that this issue perhaps should not be closed without a fix, since it creates surprising and difficult to debug breakages downstream.
